### DISCUSSION TITLE
feature: overloading std::ostream &operator<< to print all kinds of stl containers ( most widely used )

### DISCUSSION THREAD

**Discussion by Shubhambawner** [microsoft/STL#2844] [Ideas] [answered]
Hello!
one attempt I had made to overload the standard ostream<< operator that we use for printing out basic data types to console, To now be able to print out most of the data types, this will satisfy following:
0.  will support almost all majorly used containers, i.e. list, vector, set, map, pair … 
1.  will support any level of nesting, (a container inside other...)
2.  may also provide certain level of flexibility in how different collections are printed
I think the feature is worth a try, as:
0.  most of the other languages like JS already have this functionality of being able to directly log almost all kinds of things to console, keeping in view that any raw form of debugging relays on it.
1.  this would make it easy to log things in console, and so will the debugging be easy for new comers.

p.s. I am not very expert in this domain, please correct me if some better ways exist here.

Here I am appending the roughly written code, with some shots of how it runs. 

```
#include "./ru.cpp"
using namespace std;
int main()
{
    vector<vector<string>> v = {{"aa", "awa", "qaa", "aar"}, {"aa", "qaa", "aar"}, {"rtaa", "aawrawwa", "agqaa", "aahgear"}};
    vector<list<string>> vv = {{"aa", "awa", "qaa", "aar"}, {"aa", "qaa", "aar"}, {"rtaa", "aawrawwa", "agqaa", "aahgear"}};
    list<vector<int>> l = {{1, 2}, {3, 4, 5, 6}, {1, 2, 3, 4, 5, 6, 3, 4, 5, 6}, {3, 4, 5, 6}};
    list<int> s = {1, 2, 3, 4, 5, 6};
    map<int, list<vector<int>>> m = {{1, {{4, 4, 5, 6}, {6, 6}}}, {3, {{1, 4, 5}, {3, 5, 6}}}, {5, {{3, 4}, {5, 6}}}};

    cout << v << endl << endl
         << vv << endl << endl
         << l << endl << endl
         << m << endl << endl
         << s << endl << endl;
}
```
and the output:
![Screenshot (253)](https://user-images.githubusercontent.com/73178780/177413913-c5ee2c69-90e0-47e7-abcc-1730f75fd7c0.png)

(note here that we can customize how the opening or closing brackets would look)

this code for "ru.cpp" I have uploaded here:  https://github.com/Shubhambawner/learner/blob/main/DSA/recursion_DP/ru.cpp

**Comment by StephanTLavavej** [comment, marked as answer]
Implementing extensions beyond what the C++ Standard specifies is an explicit non-goal of our project - see https://github.com/microsoft/STL#non-goals .

**Comment by frederick-vs-ja** [comment]
The feature request is non-standard and non-conforming, I don't think MSVC STL should ever accept it.

In C++23 we will have Formatting Ranges ([P2286R8](https://wg21.link/P2286R8)) and Formatted output ([P2093R14](https://wg21.link/p2093r14)), so you'll be able to write `std::print(std::cout, "{}", container)` soon.
